menas/src/main/scala/za/co/absa/enceladus/rest/services/DatasetService.scala

```diff
@@ -21,19 +21,31 @@ import za.co.absa.enceladus.model.conformanceRule.ConformanceRule
 import za.co.absa.enceladus.model.{Dataset, UsedIn}
 import za.co.absa.enceladus.rest.repositories.DatasetMongoRepository
 import scala.concurrent.Future
+import za.co.absa.enceladus.model.menas.MenasReference
+import za.co.absa.enceladus.rest.models.ChangedFieldsUpdateTransformResult
+import za.co.absa.enceladus.rest.models.ChangedField
 
 @Service
-class DatasetService @Autowired()(datasetMongoRepository: DatasetMongoRepository)
-  extends VersionedModelService(datasetMongoRepository) {
+class DatasetService @Autowired() (datasetMongoRepository: DatasetMongoRepository, auditTrailService: AuditTrailService)
+  extends VersionedModelService(datasetMongoRepository, auditTrailService) {
+
+  import scala.concurrent.ExecutionContext.Implicits.global
 
   override def update(username: String, dataset: Dataset): Future[Option[Dataset]] = {
-    super.update(username, dataset.name) { latest =>
-      latest
+    super.update(username, dataset.name, dataset.version, "Dataset Updated.") { latest =>
+      val updated = latest
         .setSchemaName(dataset.schemaName)
         .setSchemaVersion(dataset.schemaVersion)
         .setHDFSPath(dataset.hdfsPath)
         .setHDFSPublishPath(dataset.hdfsPublishPath)
         .setDescription(dataset.description).asInstanceOf[Dataset]
+
+      ChangedFieldsUpdateTransformResult(updatedEntity = updated, fields = Seq(
```

Comment:
How about having a method in each entity `getUpdateChanges(updated: VersionedModel): ChangedFieldsUpdateTransformResult` which produces this for you when you call `latest.getUpdateChanges(updated)`?

Reply:
You can then use it in all create/update calls (as well as `addConformanceRule`, `updateDefaults`, `addDefault`, etc.)